### PR TITLE
split nested tests

### PR DESCRIPTION
**Pull request by wind57**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -135,6 +163,10 @@ jobs:
       - store_test_results:
           path: ~/junit/
           destination: testartifacts
+      - run:
```
Comment (wind57) [inline .circleci/config.yml:166]: output the json data that circleci uses to compute the times for tests, this is useful when debugging.

File: .circleci/config.yml [hunk 2]
```diff
@@ -37,15 +37,43 @@ jobs:
       - run:
           name: Run regular tests
           command: |
-            CLASSNAMES=$(circleci tests glob "**/src/test/**/**.java" | grep -v 'spring-cloud-kubernetes-integration-tests' \
```
Comment (wind57) [inline .circleci/config.yml:40]: use `xargs grep -L '@Nested'` to find test classes that do _not_ have `@Nested` inside them.

File: .circleci/config.yml [hunk 3]
```diff
@@ -37,15 +37,43 @@ jobs:
...
                                | sed 's/.*src.test.java.//g' | sed 's@/@.@g' \
                                | sed 's/.\{5\}$//' \
-                               | circleci tests split --split-by=timings)
-            echo $CLASSNAMES
-            TEST_ARG=$(echo $CLASSNAMES | sed 's/ /,/g')
-            echo $TEST_ARG
-            ./mvnw -s .settings.xml  -DfailIfNoTests=false -DtestsToRun=$TEST_ARG -e clean org.jacoco:jacoco-maven-plugin:prepare-agent test -U -P sonar -nsu --batch-mode -Dmaven.test.redirectTestOutputToFile=true -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn
+                               > /tmp/all_tests.txt
+
+            # find classnames that have @Nested inside them
```
Comment (wind57) [inline .circleci/config.yml:50]: the idea here is this:

we find "root" classes, that contain `@Nested` inside them; then go inside these classes and find all the inner classes that they have, basically this:

```
flattened=$(cat $p | awk '/@Nested/, /@Test/' | tr '\n' ' ' | awk '{ for (i = 1; i <= NF; ++i) { if( $i == "class") { print $(i+1) } } }')
```

once the inner classes are found, compute their fully qualified names. For example:

```
org.springframework.cloud.kubernetes.client.KubernetesClientActuatorTests$DisabledHealthTest
```

**Comment by wind57** [comment]
No matter how I look at the circleci documentation and forums and stackoverflow, their implementation of splitting tests seems buggy to me. 

It will take a few words to explain. First of all, we currently have only 7 classes that have `@Nested` inside them, these 7 generate more, of course, because I create fully qualified class names from them.  An example would be a lot easier to get, let's look at [KubernetesClientBootstrapConfigurationTests](https://github.com/spring-cloud/spring-cloud-kubernetes/blob/main/spring-cloud-kubernetes-client-config/src/test/java/org/springframework/cloud/kubernetes/client/config/KubernetesClientBootstrapConfigurationTests.java)

it has 8 `@Nested` classes. Thus I generate 8 different classes to be passed to `-DtestsToRun`. Let's track this because it is important. 

Two of them are [here](https://app.circleci.com/pipelines/github/spring-cloud/spring-cloud-kubernetes/1471/workflows/743b8078-beb8-412d-86bc-0928c75af204/jobs/2786?invite=true#step-102-55):

- `KubernetesClientBootstrapConfigurationTests$KubernetesEnabled`
- `KubernetesClientBootstrapConfigurationTests$KubernetesEnabledOnPurpose`

And [here](https://app.circleci.com/pipelines/github/spring-cloud/spring-cloud-kubernetes/1471/workflows/743b8078-beb8-412d-86bc-0928c75af204/jobs/2786/parallel-runs/1?filterBy=ALL&invite=true#step-102-55)

- `org.springframework.cloud.kubernetes.client.config.KubernetesClientBootstrapConfigurationTests$KubernetesEnabledConfigDisabled`

Three more [here](https://app.circleci.com/pipelines/github/spring-cloud/spring-cloud-kubernetes/1471/workflows/743b8078-beb8-412d-86bc-0928c75af204/jobs/2786/parallel-runs/2?filterBy=ALL&invite=true#step-102-55):

- `org.springframework.cloud.kubernetes.client.config.KubernetesClientBootstrapConfigurationTests$KubernetesEnabledSecretsDisabled`
- `org.springframework.cloud.kubernetes.client.config.KubernetesClientBootstrapConfigurationTests$KubernetesDisabled`
- `org.springframework.cloud.kubernetes.client.config.KubernetesClientBootstrapConfigurationTests$KubernetesEnabledSecretsAndConfigDisabled`

And the last [two here](https://app.circleci.com/pipelines/github/spring-cloud/spring-cloud-kubernetes/1471/workflows/743b8078-beb8-412d-86bc-0928c75af204/jobs/2786/parallel-runs/3?filterBy=ALL&invite=true#step-102-55):

- `org.springframework.cloud.kubernetes.client.config.KubernetesClientBootstrapConfigurationTests$KubernetesClientBootstrapConfigurationNotInsideK8s`
- `org.springframework.cloud.kubernetes.client.config.KubernetesClientBootstrapConfigurationTests$KubernetesClientBootstrapConfigurationNotInsideK8s`

The idea of this exercise is that I do indeed pass all the `@Nested` tests to `-DtestsToRun`.

---

Unfortunately circleci does not handle that entirely correct. When I output the json file that keeps track of all the tests that it knows about, see [here](https://app.circleci.com/pipelines/github/spring-cloud/spring-cloud-kubernetes/1471/workflows/743b8078-beb8-412d-86bc-0928c75af204/jobs/2786/parallel-runs/0?filterBy=ALL&invite=true#step-113-4) for example, only 4 of them are present. Why? I have no idea :(  

I still see entries in logs like:

```
Autodetected classname timings.
No timing found for "org.springframework.cloud.kubernetes.client.config.KubernetesClientBootstrapConfigurationTests$KubernetesClientBootstrapConfigurationInsideK8s"
No timing found for "org.springframework.cloud.kubernetes.client.config.KubernetesClientBootstrapConfigurationTests$KubernetesEnabled"
No timing found for "org.springframework.cloud.kubernetes.client.config.KubernetesClientBootstrapConfigurationTests$KubernetesEnabledConfigDisabled"
No timing found for "org.springframework.cloud.kubernetes.client.config.KubernetesClientBootstrapConfigurationTests$KubernetesEnabledSecretsDisabled
```

meaning _some_ inner tests are perfectly parsed, some are not. And why circleci does not generate these in its own json file is beyond me. Truth being said, this is not _that_ important for us. The mere fact that we have this problem, it does not mean they will not be split; they will, just not by timing. Also, because there are so little number of files that fall under this category, it's something we can totally ignore, for now. 

The other way to think about it - we are still splitting, just that circleci is not perfect about it, which is more then OK.

---

I've also opened a question about it to circleci, but my post is yet to be accepted. I will also keep an eye on this in later builds and see if this has to do with some heuristics that they have. Either way, the position we are now looks fairly good to me, still.

File: .circleci/config.yml [hunk 4]
```diff
@@ -37,19 +37,45 @@ jobs:
...
+            sed -i 's/\\$/\\\\$/' /tmp/this_node_tests_comma_separated.txt
+
+            echo '------- tests for this node ----------'
+            cat /tmp/this_node_tests_comma_separated.txt
+            echo -e '\n--------------------------------------'
+            ./mvnw -s .settings.xml  -DfailIfNoTests=false -DtestsToRun=$(cat /tmp/this_node_tests_comma_separated.txt) -e clean org.jacoco:jacoco-maven-plugin:prepare-agent test -U -P sonar -nsu --batch-mode -Dmaven.test.redirectTestOutputToFile=true -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn
             mkdir -p $HOME/artifacts/junit/
             find . -type f -regex ".*/spring-cloud-*.*/target/*.*" -exec cp {} $HOME/artifacts/ \;
             find . -type f -regex ".*/target/.*-reports/.*" -exec cp {} $HOME/artifacts/junit/ \;
-            bash <(curl -s https://codecov.io/bash)
```
Comment (wind57) [inline .circleci/config.yml:52]: this one is deprecated and their page says that:

```
"On February 1, 2022, this uploader will be fully sunset and no longer function"
```

the recommendation is to use [this one](https://github.com/codecov/uploader), but it requires `npm` and says "Heads up! This is a work in progress.". We will need to investigate a bit more on what to use and how.

**Comment by wind57** [comment]
@ryanjbaxter is there a chance for this and some other PRs to be looked at? I am _really_ not trying to be pushy in any shape or form, I am just not sure what I should be doing next. I want to contribute more and fix some of the things that I still have in my head, but I have no visibility of future roadmaps and branches, etc. :( 

As I understand some things are OK to be put in `main` (once reviewed), some, like a refactor of configmaps/secrets implementation is not (as we talked in a PR in November).  I have 6 pending PRs and about 5 more in my plans of fixing/adding more enhancements. 

As an example, I see [this roadmap for 2022.0.x](https://github.com/orgs/spring-cloud/projects/56) where there are 3 PRs of mine listed, which is lovely! But, there are no time lines, no explicit branch `2022.0.x` for in our code base created... I mean, you probably get it...

**Comment by ryanjbaxter** [comment]
Yes the PRs are on my list to get to soon.  Coming back after the holidays has been a bit insane because of the changes made on main and getting that all working (thank!) plug just digging myself out of emails etc.  Now that all the branches look stable 🤞 I can try and get back to some of the PRs and issues.

**Review by ryanjbaxter — COMMENTED** [review]
To be honest I look at how complicated this script becomes and it makes me want to just split the tests out into their own top level classes.  I would prefer to keep the circleci logic as simple as possible, especially since this stuff is so hard to test without running it on circleci.

**Comment by wind57** [comment]
oh my, this is a very valid point, I did not even think about it. Fully agree, will close this one and will re-factor the tests.

**Pull request closed without merging**